web/packages/teleterm/src/services/tshdEvents/index.ts

```diff
@@ -180,10 +185,25 @@ function createService(logger: Logger): {
         }
       );
     },
-    sendPendingHeadlessAuthentication: () => {
-      // TODO (joerger): Handle pending headless authentications with an
-      // approve/deny modal, followed by an MFA prompt for approval.
-      logger.info('Received pending headless authentication');
+    sendPendingHeadlessAuthentication: (call, callback) => {
+      const request = call.request.toObject();
+
+      logger.info('Emitting sendPendingHeadlessAuthentication', request);
+
+      const onCancelled = (callback: () => void) => {
+        call.on('cancelled', callback);
+      };
+
+      emitter
+        .emit('sendPendingHeadlessAuthentication', { request, onCancelled })
```

Comment:
A little UX thing, when you send two requests, one after the other, and then reject one of them in the UI, it seems that you still see the same, previous dialog.
This is because another request immediately comes in and displays a new dialog box. However, this happens so fast that it looks like the old dialog has not been closed.

My proposal is to wait some time before resolving the current request, so the user can see that one dialog has been closed and a new one has been opened. 

I'd do it this way:
```ts
      emitter
        .emit('sendPendingHeadlessAuthentication', { request, onCancelled })
        .then(() => wait(400))
        .then(
          () => {
            callback(null, new api.SendPendingHeadlessAuthenticationResponse());
          },
          error => {
            callback(error);
          }
        );
```    

I think we have never done something like this, WDYT @ravicious?

Reply:
Waiting is not a bad idea, but this would artificially slow down the action that requested the headless auth as well.

If anything, tshd should check how much time has passed since the last request was resolved and if it's less than x, wait a little bit before sending another request to the Electron app.

Ultimately though, I don't have a good grasp on how this feature is going to be used to warrant spending time addressing such use cases.

But the issue you've raised plays into two other things:

* Two headless requests issued for two different clusters at the same time, which I already commented on https://github.com/gravitational/teleport/pull/29097#discussion_r1275145909
   * Here we wouldn't need to wait as this would be done from two goroutines and the second one would simply not forward the request to Connect.
   * …but what if we didn't drop the requests and instead maintained a "queue" for those requests? This would boil down to using `Lock` instead of `TryLock`. We could have another field on `daemon.Service` with the time of when the lock was last released.
* The copy in the modal for two different requests is almost identical.
   * For two requests from the same cluster, we could think of showing time of the request, but even that wouldn't help much with the problem you described.

Reply:
Agreed, we can add a 1 second delay between important modal pop ups coming from the daemon. As Rafał suggested, we can do this by doing a `Lock` with a wait period afterwards if needed.

> Ultimately though, I don't have a good grasp on how this feature is going to be used to warrant spending time addressing such use cases.

I think we are getting close to edge cases with a 0% chance of occurring. Users create their own headless requests so they should not be confused by the popups. I don't know if there is a scenario where a user would need to create 2 headless requests for different clusters at the same time either.

I suppose my main concern is clarity as we don't want a user to approve their headless request, and then receive a malicious headless request, and clicking approve again thinking something went wrong with the first one.

Maybe we should make the Client IP show in larger font as it is the main thing to change if a malicious user made a request. Or we can show the request ID so they might double check it matches the request ID showing from `tsh --headless` output. For now, I'll do the latter, let me know what you think.

I don't think anything more complicated than that is warranted though.

Reply:
IMO showing Client IP in larger font would look weird, making it bold is enough.

I experimented with showing the request UUID, I hope it's no that ugly :p
<img width="644" alt="image" src="https://github.com/gravitational/teleport/assets/20583051/f426af69-a2a3-4512-aed4-1c6265f1c95d">

If this looks ok, I can commit it.

Reply:
@gzdunek Looks good to me, thanks!

Reply:
Done https://github.com/gravitational/teleport/pull/29097/commits/7ff47f3896822ae9713321c5646327856f3933db